kubeflow/mpi-job/mpi-operator.libsonnet

```diff
@@ -0,0 +1,276 @@
+          ],
+        },
+        validation: {
+          openAPIV3Schema: {
+            properties: {
+              spec: {
+                title: "The MPIJob spec",
+                description: "Either `gpus` or `replicas` should be specified, but not both",
+                oneOf: [
+                  {
+                    properties: {
+                      gpus: {
+                        title: "Total number of GPUs",
+                        description: "Valid values are 1, 2, 4, or any multiple of 8",
+                        oneOf: [
+                          {
+                            type: "integer",
+                            enum: [
+                              1,
+                              2,
+                              4,
+                            ],
+                          },
+                          {
+                            type: "integer",
+                            multipleOf: 8,
+                            minimum: 8,
+                          },
+                        ],
+                      },
+                    },
+                    required: [
+                      "gpus",
+                    ],
+                  },
+                  {
+                    properties: {
+                      replicas: {
+                        title: "Total number of replicas",
+                        description: "The GPU resource limit should be specified for each replica",
+                        type: "integer",
+                        minimum: 1,
+                      },
+                    },
+                    required: [
+                      "replicas",
+                    ],
+                  },
+                ],
+              },
+            },
+          },
+        },
+      },
+    },
+
+    clusterRole(name):: {
+      kind: "ClusterRole",
+      apiVersion: "rbac.authorization.k8s.io/v1",
+      metadata: {
+        name: name,
+      },
+      rules: [
+        {
+          apiGroups: [
+            "",
+          ],
+          resources: [
+            "configmaps",
+            "serviceaccounts",
+          ],
+          verbs: [
+            "create",
+            "list",
+            "watch",
+          ],
+        },
+        {
+          // This is needed for the launcher Role.
+          apiGroups: [
+            "",
+          ],
+          resources: [
+            "pods",
+          ],
+          verbs: [
+            "get",
+          ],
+        },
+        {
+          // This is needed for the launcher Role.
+          apiGroups: [
+            "",
+          ],
+          resources: [
+            "pods/exec",
+          ],
+          verbs: [
+            "create",
+          ],
+        },
+        {
+          apiGroups: [
+            "",
+          ],
+          resources: [
+            "events",
+          ],
+          verbs: [
+            "create",
+            "patch",
+          ],
+        },
+        {
+          apiGroups: [
```

Comment:
Can we remove rule for ```rbac.authorization.k8s.io```?
This one will give permission which equivalent to ```cluster-admin```.

Reply:
This is needed to create new roles/rolebindings to lock down the MPI jobs since they need to communicate with other pods and do remote execution. We actually create a new role/serviceaccount/rolebinding for each MPIJob. There is some information about it in the design proposal: https://github.com/kubeflow/community/blob/master/proposals/mpi-operator-proposal.md

Reply:
OK, in this case can we remove ```clusterRole``` and just do ```clusterRoleBinding``` to ```cluster-admin```?
With access to role and role binding user can get themselves admin access.
So let's save effort of define custom role.

Reply:
I think its better to have a specific role. This makes it easier to add/remove permissions needed by this role.

@rongou Do you create job specific roles/rolebindings?

Reply:
Right, the role is very specific to the individual `MPIJob`. The launcher has `pods/exec` permissions only to the worker pods. See https://github.com/kubeflow/mpi-operator/blob/master/pkg/controllers/mpi_job_controller.go#L821

Here is a sample role:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  creationTimestamp: 2018-06-11T23:14:22Z
  labels:
    app: tensorflow-benchmarks-8
  name: tensorflow-benchmarks-8-launcher
  namespace: default
  ownerReferences:
  - apiVersion: kubeflow.org/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: MPIJob
    name: tensorflow-benchmarks-8
    uid: 2baf54b7-6dcd-11e8-9b38-029ed2ab0d38
  resourceVersion: "466996"
  selfLink: /apis/rbac.authorization.k8s.io/v1/namespaces/default/roles/tensorflow-benchmarks-8-launcher
  uid: 2cf3b945-6dcd-11e8-b6ec-06d57706f108
rules:
- apiGroups:
  - ""
  resourceNames:
  - tensorflow-benchmarks-8-worker-0
  - tensorflow-benchmarks-8-worker-1
  - tensorflow-benchmarks-8-worker-2
  - tensorflow-benchmarks-8-worker-3
  - tensorflow-benchmarks-8-worker-4
  - tensorflow-benchmarks-8-worker-5
  - tensorflow-benchmarks-8-worker-6
  - tensorflow-benchmarks-8-worker-7
  resources:
  - pods
  verbs:
  - get
- apiGroups:
  - ""
  resourceNames:
  - tensorflow-benchmarks-8-worker-0
  - tensorflow-benchmarks-8-worker-1
  - tensorflow-benchmarks-8-worker-2
  - tensorflow-benchmarks-8-worker-3
  - tensorflow-benchmarks-8-worker-4
  - tensorflow-benchmarks-8-worker-5
  - tensorflow-benchmarks-8-worker-6
  - tensorflow-benchmarks-8-worker-7
  resources:
  - pods/exec
  verbs:
  - create
```